keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,661 @@
+is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
+will not be able to start. This can be mitigated by fixing the malformed values.
+
+
+## Design Details
+
+We would like to introduce a structured file format which allows authorization
+to be configured using a flag (`--authorization-config-file`) which accepts a
+path to a file on the disk. This feature can be enabled or disabled by the
+explicit feature flag `AuthorizationConfigFromFile`.
+The proposed structure is illustrated below:
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: AuthorizationConfiguration
+authorizers:
+  - type: Webhook
+    webhook:
+      # Name used to describe the webhook
+      # This is explicitly used in monitoring machinery for metrics
+      # Note:
+      #   - Do exercise caution when setting the value
+      #   - If not specified, the default would be set to ""
+      #   - If there are multiple webhooks in the authorizer chain,
+      #     this field is required
+      name: super-important-kube-system-authorizer
+      # The duration to cache 'authorized' responses from the webhook
+      # authorizer.
+      # Same as setting `--authorization-webhook-cache-authorized-ttl` flag
+      # Default: 5m0s
+      authorizedTTL: 30s
+      # The duration to cache 'authorized' responses from the webhook
+      # authorizer.
+      # Same as setting `--authorization-webhook-cache-unauthorized-ttl` flag
+      # Default: 30s
+      unauthorizedTTL: 30s
+      # Timeout for the webhook request
+      # Maximum allowed is 30s.
+      # Required, with no default.
+      timeout: 3s
+      # The API version of the authorization.k8s.io SubjectAccessReview to
+      # send to and expect from the # webhook.
+      # Same as setting `--authorization-webhook-version` flag
+      # Required, with no default
+      subjectAccessReviewVersion: v1
+      # Controls the authorization decision when a webhook request fails to
+      # complete or returns a malformed response.
+      # Valid values:
+      #   - NoOpinion: continue to subsequent authorizers to see if one of
+      #     them allows the request
+      #   - Deny: reject the request without consulting subsequent authorizers
+      # Default: NoOpinion
+      failurePolicy: Deny
+      connectionInfo:
+        # Controls how the webhook should communicate with the server.
+        # Valid values:
+        # - KubeConfig: use the file specified in kubeConfigFile to locate the
+        #   server.
+        # - InClusterConfig: use the in-cluster configuration to call the
+        #   SubjectAccessReview API hosted by kube-apiserver. This mode is not
+        #   allowed for kube-apiserver.
+        type: KubeConfig
+        # Path to KubeConfigFile for connection info
+        kubeConfigFile: /kube-system-authz-webhook.yaml
+      matchConditions:
+      - expression: |
+        request.resourceAttributes.namespace == 'kube-system'
```

Comment:
@ritazh - as a follow-up, check how this expression would behave for a non-resource request (e.g. to `/version`) where resourceAttributes is null... if this would error, we should improve this example to put a `has(request.resourceAttributes)` match condition first, etc

Reply:
I have added the conditional either way.